dask/dataframe/shuffle.py

```diff
@@ -732,7 +739,16 @@ def collect(p, part, meta, barrier_token):
 
 
 def set_partitions_pre(s, divisions):
-    partitions = divisions.searchsorted(s, side="right") - 1
+    try:
+        partitions = divisions.searchsorted(s, side="right") - 1
+    except TypeError:
+        # When `searchsorted` fails with `TypeError`, it may be
+        # caused by nulls in `s`. Try again with the null-values
+        # explicitly mapped to the first partition.
+        partitions = np.empty(len(s), dtype="int32")
+        partitions[s.isna()] = 0
+        not_null = s.notna()
```

Comment:
Do you think it'd be safer to wrap this in a try/except and raise the original error if it fails?

Reply:
We will still get a similar error if the second attempt at `searchsorted` fails, but I totally agree that we are hiding the original error if one of these three lines happens to fail (and I don't like that either).  These lines seem pretty "safe" to me - Any suggestions on whay we should be looking to catch here? (sorry If I am being slow here)

Reply:
Could we just wrap it in a 

```
try:
...
except Exception as e:
    raise e from original_e
```

After writing that out it occurred to me that that is probably already what happens, so I am fine if you prefer to leave it as is.